### PR TITLE
Add tests for POGM

### PR DESCRIPTION
**Pull request by chaithyagr**
This PR adds tests for POGM. Tracked in #42

### CODE REVIEW

File: mri/numerics/reconstruct.py [hunk 1]
```diff
@@ -443,7 +443,7 @@ def sparse_rec_pogm(gradient_op, linear_op, prox_op, cost_op=None,
         print("Done.")
         print("Execution time: ", end - start, " seconds")
         print("-" * 40)
-    x_final = opt.x_final
+    x_final = linear_op.adj_op(opt.x_final)
```
Comment (chaithyagr) [inline mri/numerics/reconstruct.py]: This is done to follow the same norm as other optimizers. We want to ensure that we return Images in the end.

Comment (zaccharieramzi) [reply mri/numerics/reconstruct.py]: I really dislike this as the `adj_op` is not necessarily an inverse (or reconstruction) operation. I don't know how it's done elsewhere, but we should let the user change the outcome the way they want.

File: mri/test/test_optimizer.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
 from mri.reconstruct.fourier import FFT2, NFFT
 from mri.numerics.reconstruct import sparse_rec_fista
 from mri.numerics.reconstruct import sparse_rec_condatvu
+from mri.numerics.reconstruct import sparse_rec_pogm
```
Comment (zaccharieramzi) [inline mri/test/test_optimizer.py]: Maybe we can have all these imports on the same line.

File: mri/test/test_optimizer.py [hunk 3]
```diff
@@ -84,6 +85,47 @@ def test_reconstruction_fista_fft2(self):
                     np.testing.assert_allclose(x_final, np.fft.ifftshift(
                         fourier_0.adj_op(data_0)))
 
+    def test_reconstruction_pogm_fft2(self):
+        """ Test all the registered transformations.
+        """
+        print("Process test FFT2 POGM::")
+        for image in self.images:
+            fourier = FFT2(samples=convert_mask_to_locations(
+                                            fftshift(self.mask)),
```
Comment (zaccharieramzi) [inline mri/test/test_optimizer.py]: Can you explain why we have to do this?

File: mri/test/test_optimizer.py [hunk 4]
```diff
@@ -84,6 +85,47 @@ def test_reconstruction_fista_fft2(self):
...
+                        prox_op=prox_op,
+                        cost_op=cost_op,
+                        max_nb_of_iter=self.nb_iter,
+                        verbose=0)
+                    fourier_0 = FFT2(samples=convert_mask_to_locations(
+                                            fftshift(self.mask)),
+                                     shape=image.shape)
+                    data_0 = fourier_0.op(np.fft.fftshift(image.data))
+                    np.testing.assert_allclose(x_final, np.fft.ifftshift(
+                        fourier_0.adj_op(data_0)))
```
Comment (zaccharieramzi) [inline mri/test/test_optimizer.py]: All these ffshifts are making me nervous. I don't understand why they are not handled directly in the FFT objects we have.

File: mri/test/test_optimizer.py (hunk 2 above)
Comment (chaithyagr) [reply mri/test/test_optimizer.py]: Well, PEP8 would still need to move to a different line, this just looks cleaner to me. Let me know if this needs to be changed still, I am fine with it also

File: mri/test/test_optimizer.py (hunk 3 above)
Comment (chaithyagr) [reply mri/test/test_optimizer.py]: Well this is something to be fixed, in another PR. FFT class right now works on shifted data, while all other Fourier classes (NFFT, NUFFT etc) works directly on mask. 
I would really like FFT to change its behaviour, but I would not do this in this PR. Maybe I will get an issue up for it. Will resolve this after referencing the issue here

File: mri/test/test_optimizer.py (hunk 4 above)
Comment (chaithyagr) [reply mri/test/test_optimizer.py]: Resolving as a duplicate of above

File: mri/test/test_optimizer.py (hunk 3 above)
Comment (zaccharieramzi) [reply mri/test/test_optimizer.py]: Yes definitely don't do it here, but it seems weird to me. I had never noticed but fair enough.

File: mri/numerics/reconstruct.py (hunk 1 above)
Comment (LElgueddari) [reply mri/numerics/reconstruct.py]: For a more general perspective, the sparse_rec_* is supposed to give the image solution to the reconstruction problem, if we assume that this is the goal of the function than we should output the image. This is for a general perspective, I think that to handle this in a correct manner, `linar_op` should have both an adjoint operator and an inverse operator, i.e. for the simplest case -- i.e. orthogonal wavelet transform, handling correctly the boundaries --  the adjoint and the inverse should be the same whereas for more complicated  transforms maybe we should implement a dedicated inverse operation.

File: mri/test/test_optimizer.py (hunk 2 above)
Comment (chaithyagr) [reply mri/test/test_optimizer.py]: Well I planned to move to that itself, just didnt know how clean it was. Will get an update

File: mri/test/test_optimizer.py (hunk 3 above)
Comment (LElgueddari) [reply mri/test/test_optimizer.py]: Most of the time when you compute your mask you are putting the DC frequency at the middle of the mask, however, the FFT operator does not compute it in the middle but rather in the boundaries, therefore you either need to add an `fftshift` operation directly on the mask or for each `FFT` operations  this was the easiest solution.

Comment (chaithyagr) [reply mri/test/test_optimizer.py]: Nah, you might have come across this even in cartesian example:

https://github.com/CEA-COSMIC/pysap-mri/blob/5bb4d2ffc95997fa180d6057ac7e3618c16e99e0/examples/cartesian_reconstruction.py#L48-L49


So there will surely be a bigger impact of this change, although it is a right change

Comment (chaithyagr) [reply mri/test/test_optimizer.py]: #45 created. Resolving here

File: mri/numerics/reconstruct.py (hunk 1 above)
Comment (chaithyagr) [reply mri/numerics/reconstruct.py]: I agree with Loubna, resolving here. We could instance an issue to obtain the reconstruction operator in both WaveletN and WaveletUD classes.

Comment (zaccharieramzi) [reply mri/numerics/reconstruct.py]: Yes sorry I was misreading, I thought this was the algorithm. This is def fine.

File: mri/test/test_optimizer.py (hunk 2 above)
Comment (zaccharieramzi) [reply mri/test/test_optimizer.py]: You didn't look at the SO answer though haha, nv mind it's ok like this

**Pull request merged**